src/main/java/org/opensearch/neuralsearch/search/query/HybridCollectorManager.java

```diff
@@ -138,16 +139,36 @@ public ReduceableSearchResult reduce(Collection<Collector> collectors) {
         }
 
         if (!hybridTopScoreDocCollectors.isEmpty()) {
-            HybridTopScoreDocCollector hybridTopScoreDocCollector = hybridTopScoreDocCollectors.stream()
-                .findFirst()
-                .orElseThrow(() -> new IllegalStateException("cannot collect results of hybrid search query"));
-            List<TopDocs> topDocs = hybridTopScoreDocCollector.topDocs();
-            TopDocs newTopDocs = getNewTopDocs(
-                getTotalHits(this.trackTotalHitsUpTo, topDocs, isSingleShard, hybridTopScoreDocCollector.getTotalHits()),
-                topDocs
-            );
-            TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(newTopDocs, hybridTopScoreDocCollector.getMaxScore());
-            return (QuerySearchResult result) -> { result.topDocs(topDocsAndMaxScore, getSortValueFormats(sortAndFormats)); };
+            List<ReduceableSearchResult> results = new ArrayList<>();
+            for (HybridTopScoreDocCollector hybridTopScoreDocCollector : hybridTopScoreDocCollectors) {
+                List<TopDocs> topDocs = hybridTopScoreDocCollector.topDocs();
+                TopDocs newTopDocs = getNewTopDocs(
+                    getTotalHits(this.trackTotalHitsUpTo, topDocs, hybridTopScoreDocCollector.getTotalHits()),
+                    topDocs
+                );
+                TopDocsAndMaxScore topDocsAndMaxScore = new TopDocsAndMaxScore(newTopDocs, hybridTopScoreDocCollector.getMaxScore());
+
+                results.add((QuerySearchResult result) -> {
+                    // this is case of first collector, query result object doesn't have any top docs set, so we can
+                    // just set new top docs without merge
+                    if (result.hasConsumedTopDocs()) {
+                        result.topDocs(topDocsAndMaxScore, getSortValueFormats(sortAndFormats));
+                        return;
+                    }
+                    // in this case top docs are already present in result, and we need to merge next result object with what we have.
+                    // if collector doesn't have any hits we can just skip it and save some cycles by not doing merge
+                    if (newTopDocs.totalHits.value == 0) {
+                        return;
+                    }
+                    // we need to do actual merge because query result and current collector both have some score hits
+                    TopDocsAndMaxScore originalTotalDocsAndHits = result.topDocs();
+                    result.topDocs(
+                        mergeTopDocsAndMaxScores(originalTotalDocsAndHits, topDocsAndMaxScore),
+                        getSortValueFormats(sortAndFormats)
```

Comment:
Can we make sure this computation getSortValueFormats is done once? Because it will not change with new collector instance. It will be same for all instances of HybridTopScoreDocCollector so just calculate it before line 142.

Reply:
ack